### ISSUE TITLE
Design error: command serials may collide

### ISSUE THREAD

**Issue by mhaberler** [machinekit/machinekit#114]
Commands passed via NML typically have integer serial number for tracking status (received, being acted upon, completed, errored).

There is a fundamental design error: serial numbers are assigned on the originating side of a command. As soon as more than one command originator is active, this creates the danger of collisions, as NML has no identifier for an originator which could be used to make a submission ID unique (using a tuple (submitter, serial)). 

An example for a hack around this issue is here: https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/emcrsh.cc#L2842

problem: specify and implement a serial scheme which guarantees uniqueness.

**Comment by mhaberler** [comment]
this is an interesting & practical [paper on timestamps in distributed systems](http://zoo.cs.yale.edu/classes/cs426/2012/lab/bib/fidge88timestamps.pdf) .

**Comment by mhaberler** [comment]
good solution, but likely overkill: [How to Have your Causality and Wall Clocks, Too by Jon Moore](https://www.youtube.com/watch?v=YqNGbvFHoKM&feature=youtu.be)

**Comment by strahlex** [comment]
Thats pretty similar to how the time synchronization of FlexRay works.

**Comment by mhaberler** [comment]
[yesterday's talk on task](https://plus.google.com/+MichaelHaberler/posts/ihqtdu3DvjR)  showed significant interest in fixing the serial number collision problem; now that the talk established some common understanding how task and UI's interact we might refocus on the problem - it is in fact more involved than just avoiding serial collisions.

this thread diverted a bit from the original problem to the mechanics of defining serials, so let me recap what the issues actually are:

1. serial numbers may collide as they are created by task clients without coordination.
- serial numbers _in the current code base_ need to be **unique** but **not monotonically increasing**
- the actual problem cannot be fixed by a guaranteed-to-be-unique serial alone
- it probably cannot be fixed as long as we have NML/RCS as middleware.

re (1):  to understand the issue, look at the code in emcmodule.cc which injects commands into task:
- [a serial number is created per-client in a non-unique way here](https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/axis/extensions/emcmodule.cc#L242-L244)
- [each new command message is tagged with this serial](https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/axis/extensions/emcmodule.cc#L863) (look for all calls to `next_serial`)
- [the command message is written to the command channel](https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/axis/extensions/emcmodule.cc#L865)
- the [emcWaitCommandReceived function](
https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/axis/extensions/emcmodule.cc#L230-L240) polls the status channel for an update of EMC_STAT by task, which would contain the [echo_serial_number](https://github.com/machinekit/machinekit/blob/master/src/libnml/nml/stat_msg.hh#L24) which was filled in by task to acknowledge reception of the command
- the [echo_serial_number is tested for equality](https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/axis/extensions/emcmodule.cc#L235) - if yes, the client assumes the command has been received (btw with any error checking - emcWaitCommandReceived is a void function)
- the [same flow is applied to test for an injected command being complete](https://github.com/machinekit/machinekit/blob/master/src/emc/usr_intf/axis/extensions/emcmodule.cc#L212-L228) - note the comparison of status for RCS_DONE and RCS_ERROR both of which signal completion - RCS_EXEC would signify _command still in progress_.

re (2): I mentioned I was not sure about serial numbers being assumed to be monotonically increasing - this is not the case. To see why, one can grep for all uses of `echo_serial_number` in the code base and check which comparison operators are applied. This command does that: `grep -r echo_serial_number src/emc/|egrep -e '(>|<)'` - the hits contain no use of `<`  or `>` operators. **However, the patch by @sittner referenced in [linuxcnc bug 328](https://sourceforge.net/p/emc/bugs/328/) would change that** to `monotonically increasing serials are assumed` . A variant thereof has found its way into the linuxcnc code base, which partially fixes the problem. 

re (3): **unique serials are not sufficient.** A second problem is the semantics of the NML status channel - **it does not provide queued communication but rather has a `last update counts` semantics**. @sittner [describes the issue here](https://sourceforge.net/p/emc/bugs/328/#9a3d): if several clients are injecting commands, the EMC_STAT structure as provided by the status channel **only shows the last update**, a previous update which - for whatever reasons, like random delays - has not been received in time by the originating client **is lost** as it is overwritten by a faster second client. So this is a **race condition** between several clients updating their view of EMC_STAT and task updating same.

To summarize, the "global shared buffer update" as provided by the NML status channel **does not provide the semantics necessary to funnel individual acknowledgements to their respective originators**. 

**What would be needed is an RPC-style per-client bidirectional communications pipe** which has the following property:
1. a command injected is tagged with a unique originator id.
2. task must be able to monitor an arbitrary number of such per-client pipes.
3. Responses to a request coming in on a particular pipe must be directed _only_ to the originating client, based on the originator id from (1).
4. These pipes must support queued semantics - no update may get lost either way.

(1) assures task cannot get confused where a command came from.
(2) assures several UI's may work in parallel.
(3) and (4) fix the _lost update_ problem.

I note that NML does not support such an RPC-style pattern. I have tried in the past, and it is essentially not doable for the lack of queued operations, private pipes between using entities, and a lack of a originator ID mechanism which aids the identification of the originator and the routing of the replies. **So frankly this might not be doable at all properly as long as we use NML.**

Note that zeroMQ has all what is needed - a DEALER socket in a client and a ROUTER socket in task does provide the required semantics. zeroMQ also takes care of unique client ID's so the tuple of (client ID, serial) would still be unique. That said, it probably would still be desirable to have unique ID's because it aids tracking messages as the flow through the system, and they would be more HAL-friendly than a tuple.

So, if one were to go about to fix this design flaw once and for all:

- the _duplicate serials_ is actually not the core issue.
- trying to fix this flaw within the existing NML framework is a dead end.
- using zeroMQ in the command path will fix the issue for good. One option is to migrate selectively - introduce zeroMQ into task, but just have it work on command input for now while retaining the NML message format but dropping the unsuitable RCS communications path.

On a migration strategy:

I actually did some work addressing the last item but that was never merged. However, it worked as advertised, and mixing zeroMQ transport and NML message contents was not an issue. 

The problem is rather isolated - the command input really. What we need is some impact analysis - need to identify all code which injects command into task, and see if they all are needed. 

````
$ grep -rl echo_serial_number emc/
emc/usr_intf/emcsh.cc
emc/usr_intf/emclcd.cc
emc/usr_intf/emccontroller.cc
emc/usr_intf/shcom.cc
emc/usr_intf/keystick.cc
emc/usr_intf/emcrsh.cc
emc/usr_intf/axis/extensions/emcmodule.cc
emc/usr_intf/halui.cc
emc/usr_intf/xemc.cc
emc/usr_intf/schedrmt.cc
emc/iotask/ioControl.cc
emc/iotask/ioControl_v2.cc
emc/task/emctaskmain.cc
emc/task/taskintf.cc
emc/task/taskclass.cc
emc/task/iotaskintf.cc
`````
Out of these, **emcsh emclcd keystick emcrsh emcmodule halui xemc schedrmt** are bona-fide UI's so those would have to migrate in concert; some might have to go anyway like schedrmt

thinking of it - there might be a way to do this piecemeal:

- leave the current input channel in place so the above UI's keep working
- add a parallel input channel based on zeroMQ/NML which is just an alternative way of talking to task; the old UI's would not be affected
- then migrate emcmodule.cc (the 'import linuxcnc' extension) - this would use the new channel exclusively
- this should actually fix the issue for all code which uses 'import linuxcnc' - which is most of the usage anyway
- once that code path is known to be stable, the other UI's could be migrated at leasure, taking clues from emcmodule

**Comment by ArcEye** [comment]
> I actually did some work addressing the last item but that was never merged. However, it worked as advertised, and mixing zeroMQ transport and NML message contents was not an issue. 

Can you point to this as a starter, in your repo or wherever.

**Comment by mhaberler** [comment]
this is the (branches - for-sascha, and the underlying zmq-submit-task.commands)  http://git.mah.priv.at/gitweb?p=emc2-dev.git;a=shortlog;h=refs/heads/for-sascha

this was about using the 'cheesestand ticket' analogy I referred to, so that is just one option

when reading this branch, I suggest to concentrate on the diffs of src/emc/task/emctaskmain.cc ("server" side) and src/emc/user_intf/axis/extensions/emcmodule.cc ("client" side)

I did [some writeup on migration](http://git.mah.priv.at/gitweb?p=emc2-dev.git;a=blob;f=src/emc/task/nmlzmq.txt;h=df13d053ed100dd28a851f57059b9df6b3ed2918;hb=3e2587d3521031bfe518f922407c948bb6ea0fab) back then, but that was about taking too big a bite at one step (including protobuf and whatnot) - take with a big lump of salt

Also since @einstine909 is now on the canon case, we will have several options downstream

But IMO just adding a zeroMQ path in parallel to the NML input channel looks perfectly doable and pretty well isolated

**Comment by ArcEye** [comment]
I suspected you had already 'invented the wheel', just needed a tyre change :smile: 
Won't be able to look properly for a few days.  

emcmodule.cc doesn't do anything for me, but I thought of implementing the server in emctask as a parallel command insertion route and modifying a copy of my C++ access libraries to pump their commands by that route.

If that worked, someone would just need to convert perfectly good C++ code into parsel-tongue bindings, to get the python equivalent :snake:

**Comment by mhaberler** [comment]
oh, let me have a look in the evening, I'll clean up and rebase, and review what I _actually_ did (and what _actually_ worked)

it could be just that the standalone injection demo was what worked, and I had not touched emcmodule yet at all

**Comment by bobvanderlinden** [comment]
@mhaberler Saw your talk yesterday, thanks for all the information!

Avoiding the duplicate serials globally is a good idea. Regarding using ZeroMQ, does that mean `milltask` is being changed to have something like http://zguide.zeromq.org/page:all#Shared-Queue-DEALER-and-ROUTER-sockets (Figure 16)?

**Comment by klerman** [comment]
@mhaberler 

Might it be possible to write a wrapper around zeromq to make it have the NML interface and just drop it in everywhere? Then migrate code to use zeromq directly as appropriate (where additional functionality is needed)?

Ken

**Comment by mhaberler** [comment]
@bobvanderlinden yes, the DEALER (client) - ROUTER (task) combination is the best match and is what I used in my experiment back then for command injection

for the back channel (task->clients) there are two options:

1. just use the same socket for replies - reliable but reply only visible to client; pretty much an RPC pattern
- use a PUB socket for general task status reporting (including all of io/motion/aux/ etc status blobs, reported as distinct topics eventually), and use that as return channel for command acks (received/in progress/done as yet one more topic)

the latter has the advantage of all status going through a single channel so no proliferation of special-purpose channels (for instance the error channel is another candidate to collapse into general status publishing); and it is also visibly to any interested party (debugging, UI's)

the downside is: there are really two kinds of status information. One is: last value counts - here you can afford to loose an update, since the next update will correct things. The other kind is: every value counts - and here one cannot afford to loose an update; eg if serial matching is employed and updates might be lost, then the client would deadlock waiting for a lost update. 

The likelihood of that happening is extremely low and there might be ways around it, but it is a concern at least in theory; the drop might happen sender side (task) under memory pressure. Fact is - most of the other content in EMC_STAT which will be funneled through a PUB socket is 'last value counts' so no issue there.

I have not made up my mind for good which is the preferable pattern; it might also be possible to reply on both channels (pub for the casual observer, and DEALER reply channel for the command injector) - actually I'd appreciate a comment by @strahlex (what do you think?)

now for the experiment back then which I'm currently rebasing onto machinekit master,
for the back channel (task acks -> client) I used a PUB/SUB pattern and that worked fine.

**Comment by mhaberler** [comment]
@klerman - wrappers usually sound like great migration aids

the reasons why I think they do not significantly aid migration in this case are: the "NML interface" has two parts: the message encoding, and the command/status/error channels.

Fact is: the channel abstraction is broken - status kind of works, but the error channel is dismal and command injection as deployed is broken beyond repair, even if you made serials unique (see further down).

So what do you wrap - water down zeroMQ flows to the extent they match that channel pattern in all brokenness, to get around  the "need to change source" issue? The result will be functionally not better than what we have. You cannot fix the serial issue by clever macros - the flows are defunct.

The other migration issue which cannot be fixed by the clever wrapper: there are no non-C++ bindings to NML encoding. Real boat anchor, say for rewriting task in Python.

I think the better strategy is:

- introduce zeroMQ patterns for interaction piecemeal, for instance for now for the command channel.
- do that in _parallel_ to the legacy interaction, which is clearly possible.
- initially use NML in the new patterns, then move on to using protobuf encoding for that channel.
- make emcmodule.cc to work with both - default to new style, and have it fall back to the legacy method in case of problems, eg through an environment variable, for the sake of testing and big isolation.
- move that partial migration into master. Give it a shakeout by users.
- 80/20: always concentrate on the linuxcnc module (emcodule.cc) because that covers most of the UI needs. I frankly do not care about keystick or mini using zeroMQ right away (they will keep working on the old channels).

There is another fundamental issue with multiple UI's injecting commands: while it might be possible to inject commands into task from several UI's, it is unclear how they should be processed - sequenced, or in parallel, or sequenced plus priority. 

Example:

Run an MDI command which takes some time, like a feed move. A second UI comes in and also executes a feed move. One would think - sequenced behavior of treatment of commands is the right thing to do - run move 2 after move 1 completes.

Now replace the second move by an abort command. The result is not what you want: the abort treated **after** move 1 completes, by which time the tool might have banged into something.

It looks task was not designed with this scenario in mind; people kind of stumbled onto the fact that one could inject commands from parallel as the channel API permits that, but not much thought was given what the execution semantics should be and if this is actually a safe thing to do.

A possible solution to that is to distinguish commands into "queueable" and immediate commands, the latter taking effect immediately (and possibly affecting the queue, eg by flushing it, like in the case of abort).

**Comment by ArcEye** [comment]
Now looked at your original code - you have pretty much already done it. :smile: 

I'll wait for your merge into a current branch, it currently fails to build with protobuf errors that I recall from quite a while back. 
( protobuf/generated/types.pb.h not being created etc - not sure if current libs are compatible )

**Comment by mhaberler** [comment]
(trying to _really_ understand what I was talking about last time ;)

Some notes on the task state machine
-------------------------------


The task execution state ('execState') is one of the following:
````
// types for EMC_TASK execState
enum EMC_TASK_EXEC_ENUM {
    EMC_TASK_EXEC_ERROR = 1,
    EMC_TASK_EXEC_DONE = 2,
    EMC_TASK_EXEC_WAITING_FOR_MOTION = 3,
    EMC_TASK_EXEC_WAITING_FOR_MOTION_QUEUE = 4,
    EMC_TASK_EXEC_WAITING_FOR_IO = 5,
    EMC_TASK_EXEC_WAITING_FOR_MOTION_AND_IO = 7,
    EMC_TASK_EXEC_WAITING_FOR_DELAY = 8,
    EMC_TASK_EXEC_WAITING_FOR_SYSTEM_CMD = 9,
    EMC_TASK_EXEC_WAITING_FOR_SPINDLE_ORIENTED = 10
};
````
The starting, and completion state is of the task state machine is EXEC_DONE.

execState is relevant only for executing commands off the canon queue
(interplist) - a new command will be fetched ONLY if the state is
EXEC_DONE.

execState is NOT relevant for 'immediate' commands (those
passed by a UI through the input channel).

However, several immediate commands cause canon commands to be
emitted, either by calling upon the interpreter, or by directly
queuing commands through emcTaskQueueCommand.


Each cycle of task executes emcTaskPlan() and emcTaskExecute() in turn.

emcTaskPlan() reads any new command (emcTaskCommand) from the input channel, and
decides what to do with it based on the mode (manual, auto, mdi) or
state (estop, on) of the machine. Many of the commands just go out
immediately to the subsystems (motion and IO). In auto mode, the
interpreter is called and as a result the interp_list is appended with
NML commands. emcTaskCommand is the 'pending command' (currently
operated upon); when done, it is zeroed.

emcTaskExecute() executes a big switch on execState.  if the status is
EXEC_DONE, it gets the next item off the interp_list, and sets
execState to the preconditions for that. These preconditions include
waiting for motion, waiting for IO, etc. Once they are satisfied, it
issues the command, sets execState to the postconditions, and zeroes
the pending command. Once the postconditions are satisfied, it gets
the next item off the interp_list, and so on.

preconditions and postconditions are only looked at in conjunction
with commands on the interp_list. Immediate commands won't have any
pre- or postconditions associated with them looked at.

Single-stepping is handled in checkPreconditions() as the first
condition. If we're in single-stepping mode, as indicated by the
variable 'stepping', we set the state to waiting-for-step. This
polls on the variable 'steppingWait' which is reset to zero when a
step command is received, and set to one when the command is
issued.

Any execution error sets execState to EXEC_ERROR. On the next
emcTaskExecute(), in the EXEC_ERROR state all queues are flushed,
motion, interpreter,spindle, iocontrol, stepping etc is stopped and
execState is set to EXEC_DONE. A sync command is queued so next cycle
the interpreter is synced.

**Comment by mhaberler** [comment]
I've rebased the 'inject task commands via zeroMQ' proof-of-concept onto machinekit master, it is here: https://github.com/mhaberler/machinekit/commits/task-zmq-cmd-channel

this implements [this ticket pattern](http://static.mah.priv.at/public/ticket-pattern.png) - you might want to read up on the tracker issue and the discussion with Sascha back then to get the idea

have a look at src/emc/task/README.zeromq to get the gist

**Comment by ArcEye** [comment]
It certainly works.

If you turn on ZDEBUG you can trace the Axis calls that use emcmodule.cc ( eg machine on ) and see them being passed through the zmq mechanism

```
[011] emcmod25012
[029] 08139A0118F90100000000000018000000000000000000000003000000
task: received EMC_TASK_SET_STATE ticket 24
15-12-16 12:28:08 task: reply message:
--------------------------------------
[011] emcmod25012
[013] 08900372085518000000A00104
emcTaskPlan new command EMC_TASK_SET_STATE, 24/0
Issuing EMC_TASK_SET_STATE --  (+505,+24,   +24,    +3,)
emcmod25012: got ticket = 24
NML_INTERP_LIST::append(nml_msg_ptr{size=24,type=EMC_TASK_PLAN_SYNCH}) : list_size=1, line_number=243
emcTaskPlanSynch() returned 0
----- current state: EXEC_DONE: precondition   EXEC_WAITING_FOR_MOTION_AND_IO for command EMC_TASK_PLAN_SYNCH
NML_INTERP_LIST::append(nml_msg_ptr{size=24,type=EMC_TASK_PLAN_SYNCH}) : list_size=1, line_number=243
task: update ticket 24 status=RCS_RECEIVED->RCS_EXEC to origin=emcmod25012
listener emcmod25012: cticket = 24 status=2
----- current state: EXEC_DONE: precondition   EXEC_WAITING_FOR_MOTION_AND_IO for command EMC_TASK_PLAN_SYNCH
----- transition (io+motion done) EXEC_WAITING_FOR_MOTION_AND_IO  -> EXEC_DONE
Issuing EMC_TASK_PLAN_SYNCH --  (+516,+24,    +0,)
emcTaskPlanSynch() returned 0
----- current state: EXEC_DONE: postcondition   EXEC_DONE for command EMC_TASK_PLAN_SYNCH
task: update ticket 24 status=RCS_EXEC->RCS_DONE to origin=emcmod25012
listener emcmod25012: cticket = 24 status=1
```

Will take a while to get my head around it.

In the interim, the true awfulness of the alleged C++ legacy code in task, hits you in the face.

Looks like someone thought it would be a 'clever' idea to type everything as 
`class EMC_XXX : public RCS_CMD_MSG........`
so that you could use inheritance, instantiate it and use its methods / access it data members
but that is as far as any C++ code extends. 
The .cc files contain just C functions and are littered with global variables:-1:

**Comment by mhaberler** [comment]
I discussed with @strahlex patterns yesterday and I'll use the "commands and status updates both via the dealer/router channel" scheme, working on it

that said, it occurred to me that there is a design flaw with respect to "task execStatus"

as far as a UI is concerned, it is not particularly interesting what the "task execStatus" is, whatever that means - it is more interesting to see the status of a particular _command_ a UI just posted

now as long as there is only one input channel (NIST design singletonitis again) it does not matter where that single status value lives; as soon as there can be several UI's injecting commands there is no way to express the status of a particular command, which is where task.execStatus gets pretty useless and status updates of individual commands go haywire, which I guess is part of the problems observed

what is needed therefore is a representation in task of _all commands_ currently pending, not just the last one, and a means to associate status with a particular command - only then all of them can be properly closed off in case of say an EMC_TASK_ABORT and with the right vehicle (update echo_serial_number? report status on dealer socket? one needs the origin to do the right thing). With only the last command being represented one is bound to miss something eventually.

what is also lacking is a basic liveness test of the UI/task interaction: run axis, kill -9 milltask and notice the UI has no clue that milltask went away - it just freezes. So a periodic ping scheme is on order.

that aside, command injection done properly with zeroMQ will be a great wedge to drive in along the 'dump NML' caravan path, since it can live perfectly fine in parallel with legacy mode and switched as needed in case of problems

so as an outline what I'll do on the issue:
- I'll finish the prototype ticket pattern scheme so emcmodule.cc works with command injection and command execution status just over a dealer socket (no separate status publisher/subscriber pair)
- once this works again, emc/task gets a brushup: the monster source files will be split up into per-function files and includes to make coupling explicit, plus fixing some of the most blatant uses of globals by replacing them with parameters - this refactor will happen before any zeroMQ stuff is applied and should go straight into master since no functional change
- re-applying the zeroMQ injection will go into a development branch thereafter
- we need to run the task cycle off a czmq event loop (zloop), the reason being: we need to announce the zeromq sockets via mDNS and that avahi integration requires a zloop inner loop - nothing major, mostly a rehsuffle
-  next is the 'pending command representation' issue - likely a list or dict of structs recording origin, command and respective status
- then zmqcmd.cc and emcmodule.cc fixed up to work with this scheme
- then we can look into merging

**Comment by ArcEye** [comment]
> next is the 'pending command representation' issue - likely a list or dict of structs recording origin, command and respective status

This is my starting point for a linked list, using the utlist headers (http://troydhanson.github.io/uthash/utlist.html )
https://github.com/ArcEye/machinekit/commit/f78444673a0b89a2087238b7271877c0bb74ff55

Currently just creates a list and prints out contents at task exit.

The other option is using a hash based on the ticket number, which would be a good key as it should be unique (whole point of all this :laughing:  )

What do you think?

**Comment by ArcEye** [comment]
Now moved into zmqsupport.cc and state updated in the  publish_ticket_update() loop

Full output has updated states of tickets.
```
14 elements in cmd_list
From emcmod20453 - Ticket No. 2 - Command EMC_TASK_PLAN_SET_BLOCK_DELETE - Status = RCS_DONE
From emcmod20453 - Ticket No. 3 - Command EMC_TASK_PLAN_SET_OPTIONAL_STOP - Status = RCS_DONE
From emcmod20453 - Ticket No. 4 - Command EMC_TASK_SET_MODE - Status = RCS_DONE
From emcmod20453 - Ticket No. 5 - Command EMC_TASK_SET_MODE - Status = RCS_DONE
From emcmod20453 - Ticket No. 6 - Command EMC_TASK_PLAN_OPEN - Status = RCS_DONE
From emcmod20453 - Ticket No. 7 - Command EMC_TRAJ_SET_SCALE - Status = RCS_DONE
From emcmod20453 - Ticket No. 8 - Command EMC_TRAJ_SET_SPINDLE_SCALE - Status = RCS_DONE
From emcmod20453 - Ticket No. 9 - Command EMC_TRAJ_SET_MAX_VELOCITY - Status = RCS_DONE
From emcmod20453 - Ticket No. 10 - Command EMC_TRAJ_SET_SPINDLE_SCALE - Status = RCS_DONE
From emcmod20453 - Ticket No. 11 - Command EMC_TRAJ_SET_SCALE - Status = RCS_DONE
From emcmod20453 - Ticket No. 12 - Command EMC_TASK_SET_STATE - Status = RCS_DONE
From emcmod20453 - Ticket No. 13 - Command EMC_TASK_SET_STATE - Status = RCS_EXEC
From emcmod20453 - Ticket No. 14 - Command EMC_TASK_SET_MODE - Status = RCS_DONE
From emcmod20453 - Ticket No. 15 - Command EMC_TRAJ_SET_TELEOP_ENABLE - Status = RCS_DONE
```

Needs a polish but seems entirely workable.
Repo updated

If you can work with this method, I will expand upon it.

**Comment by mhaberler** [comment]
hold those horses ;) not clear yet what the datastructure needed is since we have not fully thought through the problem yet; thinking aloud now:

1. Acknowledgments - legacy NML. Simple if just one UI legacy channel. Just sets the echo_serial_number and status to whatever is being completed and write the status buffer. Broken if more than one UI connected, likely since serials go haywire in EMC_STAT.

2. Acks - zeroMQ and NML in parallel: different ack mechanisms; on zeromq/ticket scheme, immediately replies with assigned ticket/RCS_RECEIVED then RCS_EXEC/RCS_DONE/RCS_ERROR acks as results come in. Note that we need to record the nature and origin of the current command so we can do the right thing according to channel - bump echo_serial.. on NML or reply on the _right_ dealer channel (assuming the general case - there are several zeromq UI's).

3. A pending command is aborted by either input channel receiving an EMC_TASK_ABORT. Now both need acks (they could even come from different channels). That is currently not possible since we only record the last command and origin (which would be the abort, so the pending command being aborted really goes unacknowledged; we do not want that to avoid a UI deadlocking waiting for a serial/ticket in the ack which never shows up.

To address the above three, we need to record the _currently pending  command_ (but not the whole sequence) - and in case the abort comes in on the same channel as the pending command, the abort origin/type as well. On an abort we iterate the pending commands in ascending serial/ticket per channel, close those off, then ack the abort.

On top of this we have the current have-baked "MDI Queuing" behavior (looong thread back then on emc-developers). Here is the scenario:

- example - you fire off two MDI commands _without_ waiting for RCS_DONE after the first one.
- what happens is - both MDI commands are passed to the interpreter right away, and populate the canon queue (the mythical "MDI queue" is in fact the canon queue).
- this just happens to work if the MDI commands have NO queue buster ops. If the first one has a queue buster, but the second MDI has not, the execution of the second MDI happily tramples over the interpreter state of the first command (which might still be waiting for the queue to finish, execute the queue buster and continue
- the whole thing is a bug redressed as a feature. it happens because task blindly calls the interpreter read/exec functions without regard what is still going on.

Just saying we need to keep this one in mind. The more I think of it though - I guess #106 will take care of this as the interpreter becomes master of its main loop, and will be fed through a queue (another dealer/router pair) between task and interpreter - it will just not start executing the second MDI before the first one finishes. So then we'll have a proper queueing behavior of the interpreter as a result of closing off #106. Once the interpreter dequeues a command she'll ack the command, and task can send an RCS_EXEC update to the originator.

Anyway - the simplest and most general tracking mechanism is likely a list. This is how it would work:

- every command coming in is recorded in this list, by appending, with: channel, serial or ticket, status.
- status change of a ticket or serial: search list for match of serial/ticket and channel. Record status; or if status is RCS_DONE, delete the list item after sending update in proper channel/mechanism.
- abort: add the abort command. Execute it. when done, complete the whole list of pending commands starting oldest first, acking with proper channel/mechanism.

To verify we got things right, a periodic scan on task idle must meet an empty command list. If not, we must have missed an ack which is an error.

**Comment by mhaberler** [comment]
@ArcEye your code seems pretty much there; we do need NML channel commands just alike 

If I have a say I would go for a std::list instead of the C macros you proposed - much harder to screw up ;)

not sure how to handle the origin ID in case of several NML UI's present - there _is_ no originator id

but then we could skip the issue and assume the multiple-UI case is fixed ONLY in the zeromq scenario, and for NML only one channel is supported - no point in saving the whales now

**Comment by ArcEye** [comment]
> If I have a say I would go for a std::list instead of the C macros you proposed - much harder to screw up ;)

Implemented:  ArcEye/machinekit@2277eb4
It is just a place holder for now really, it is the connections and usages that need the thought.

> not sure how to handle the origin ID in case of several NML UI's present - there is no originator id
but then we could skip the issue and assume the multiple-UI case is fixed ONLY in the zeromq scenario, and for NML only one channel is supported - no point in saving the whales now

I think that is the only sane way to go.
The longer term aim is to get rid of NML, if the emcmodule.cc uses zmq to transmit the commands, that takes care of 90% of cases via Axis and gladevcp panels, which both use the python bindings.
Not sure about Alex's remote stuff though.

I am sure I could break it using my C++ libraries, but since I pretty much only do single user, single UI, I would have to create an extra active panel that uses NML commands directly to do that.

**Comment by mhaberler** [comment]
Alex's mkwrapper sits on top of the current emcmodule and is written in Python, so no impact

Post having gotten rid of NML, all of emcmodule.cc can be rewritten in plain Python (using import zeromq protobuf etc), and we can get rid of the polling and use proper event notification

**Comment by ArcEye** [comment]
> status change of a ticket or serial: search list for match of serial/ticket and channel. Record status; or if status is RCS_DONE, delete the list item after sending update in proper channel/mechanism.

Now implemented in the same  publish_ticket_update() section that updates the state
If state update is RCS_DONE, deletes that element.

Also added a check, just for debug purposes, at the end of the main() loop in task, that checks if 
EMC_TASK_INTERP_IDLE is set, and if so whether there are uncompleted commands.

In theory this should never get above a count of 1 in normal use, (command to be executed on next iteration of the loop) but not properly tested yet with abort midway through running program etc.

https://github.com/ArcEye/machinekit/commit/8a84ab5239a5a21ab841f3229a6c2fffe84ada7e

**Comment by mhaberler** [comment]
taking the liberty to record @koppi's suggestion:

>Von: Jakob Flierl <jakob.flierl@gmail.com>
Betreff: Aw: [Machinekit] Free CAD / CAM / CNC combination | OpenJSCAD + Kiri:Moto + Machinekit
Datum: 24. Juni 2016 um 00:21:31 MESZ
An: Michael Haberler <mail17@mah.priv.at>
Kopie: Machinekit <machinekit@googlegroups.com>

>https://github.com/willemt/ticketd is a C implementation of a
distributed durable unique 64bit ID server on top of
https://github.com/willemt/raft which has it's theoretical roots on
Paxos but is more practical for implementation ( – most Paxos protocol
implementations in fact look more similar to the
https://en.wikipedia.org/wiki/Raft_(computer_science) protocol).

>This could be useful, if you want to synchronize IDs among tasks of
two distributed Machinekit instances. But for the simple use-case
above, the techinque with allocating serial ranges is probably a good
choice as a trade-off between implementation complexity and feature
completeness.
